Allow hotkey selection for actors owned by multiple players

dragunoff opened the pull request:
~~The first commit adds the ability to select actors of the same type(class) owned by your allies (and by any player in the case of spectators). It works by double clicking on an actor or by using the hotkey with selected actor(s).~~

The ~~second commit~~ PR adds the ability to use selection hotkeys for multiple players. This feature is only relevant in spectator mode and can be useful for keeping track of the action (now that we have target lines for spectators).

Pressing `q` OR `w` will select units of the current render player or for everyone if we are spectating as "Everyone" or with shroud off. 

~~Pressing `w` will select all units of the same type based on the currently selected units but it will work for multiple owners~~. 

Closes #16044

teinarss commented:
Something is off with the q key. I did some testing in a team game with some ai players. After a while surrender to test as a spectator, when pressing q as a spectator it only selected units from one of my previous allies.

dragunoff commented:
@teinarss I can't test that now but I think it's because selection priority is given to the render player. So if you are spectating from the POV of one of your former allies then the `q` hotkey will select only their units. If you switch to the vision of another player then selection priority goes there. 

I wonder if that should be changed in the case of team game spectators as there is actually no real "team view" - it's always from the perspective of one of the team members. I'd say we should leave it like it is as that's how the game works though I've already introduced a kind of exception here for "Everyone/Shroud off".

teinarss commented:
The problem was that i was spectating with shroud off/everyone.

dragunoff commented:
@teinarss That was good catch. I have amended the code to account for that case.

abcdefg30 commented:
> OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:L251: [SA1408] Insert parenthesis within the conditional AND and OR expressions to declare the operator precedence.

dragunoff commented:
@abcdefg30 Fixed. I gotta get used to running `make check`.

pchote commented on OpenRA.Game/SelectableExts.cs:
```diff
@@ -31,7 +31,7 @@ public static int SelectionPriority(this ActorInfo a)
 		{
 			var basePriority = a.Info.TraitInfo<SelectableInfo>().Priority;
-			var lp = a.World.LocalPlayer;
+			var lp = a.World.RenderPlayer;
```
Please change this to `var viewer = a.World.RenderPlayer ?? a.World.LocalPlayer`,  and then update the `lp` references below.  The fallback to LocalPlayer is needed for players who enable the visibility cheat (which sets RenderPlayer null).

dragunoff replied:
The fallback had to be a little different:
```
var viewer = World.LocalPlayer.Spectating ? World.RenderPlayer : World.LocalPlayer;
```

That's because `RenderPlayer` is `null` when spectating in "Disable shroud" mode and the fallback in that case was breaking the feature.

dragunoff replied:
It actually had to be:
```
var viewer = (World.LocalPlayer == null || World.LocalPlayer.Spectating) ? World.RenderPlayer : World.LocalPlayer;
```

To account for spectators that are not players.

pchote replied:
I'm confused by
> That's because `RenderPlayer` is `null` when spectating in "Disable shroud" mode and the fallback in that case was breaking the feature. 

`a.World.RenderPlayer ?? a.World.LocalPlayer` means "RenderPlayer, unless its null, in which case use LocalPlayer". Spectators in the "Disable shroud" mode should have both `RenderPlayer` and `LocalPlayer` null, so it shouldn't matter which of the two nulls it uses?

We use this pattern in 13 other places already, so if there is a case where this fails then this will need to be fixed in all of them (but not here!).

dragunoff replied:
Spectators that used to be players have only `RenderPlayer` as `null` while `LocalPlayer` is still the player they used to be while playing.

dragunoff replied:
I mean - the fallback to `LocalPlayer` was breaking the feature in "Disable Shroud" mode for players that have resigned. I don't know if that's a problem in the other places where this pattern is used.

dragunoff replied:
`var viewer = a.World.RenderPlayer ?? a.World.LocalPlayer` does not do the trick here. I've tried and it breaks the feature for defeated players that are spectating because `LocalPlayer` in that case is not null.

pchote commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -130,13 +130,13 @@ public override bool HandleMouseInput(MouseInput mi)
 							.WithHighestSelectionPriority(mousePos);
 
-						if (unit != null && unit.Owner == (World.RenderPlayer ?? World.LocalPlayer))
+						if (unit != null && unit.Owner.IsAlliedWith(World.RenderPlayer))
```
Likewise here, the fallback is needed for players with the visibility cheat enabled.

pchote commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -247,22 +247,23 @@ public override string GetCursor(int2 screenPos)
 		public override bool HandleKeyPress(KeyInput e)
 		{
-			var player = World.RenderPlayer ?? World.LocalPlayer;
+			var localPlayer = World.RenderPlayer;
```
And here

pchote commented:
Changing things related to player visibility has always been risky, so considering that this is fixing a long-standing limitation rather than a recent regression its better to leave this to Next + 1. Not worth the risk of delaying the final release this cycle considering the timing of the RAGL.

dragunoff commented:
@pchote I agree, let's defer this to Next+1.

dragunoff commented:
Rebased.

teinarss commented:
W - select across map dont seam to work as spectator?

teinarss commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -270,22 +272,38 @@ public override bool HandleKeyPress(KeyInput e)
 						return false;
 
+					var selectedActors = World.Selection.Actors
+						.Where(x => !x.IsDead);
```
```suggestion
						.Where(x => !x.IsDead)
						.ToList();
```

teinarss commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -270,22 +272,38 @@ public override bool HandleKeyPress(KeyInput e)
+					if (!owners.Any())
+						return false;
+
+					var ownedActors = selectedActors.Where(x => owners.Contains(x.Owner));
```
```suggestion
					var ownedActors = selectedActors.Where(x => owners.Contains(x.Owner)).ToList();
```

teinarss commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -270,22 +272,38 @@ public override bool HandleKeyPress(KeyInput e)
+						.Select(a => a.Owner)
+						.ToList()
+						.Distinct()
+						.Where(o => o.IsAlliedWith(viewer));
```
```suggestion
					var owners = selectedActors
						.Select(a => a.Owner)
						.Distinct()
						.Where(o => o.IsAlliedWith(viewer))
						ToList();
```

teinarss reviewed (changes requested):
Some suggestions to prevent multiple enumerations.

dragunoff committed:
Add support for "select all" and "select by type" for multiple players

dragunoff commented:
Rebased and streamlined this a little. 

"Select by type" now works for all players, not only for the players in the original selection. And it doesn't work for allied players. So this feature only changes how things work for spectators (both in "Disable shroud" and "All players" mode).

abcdefg30 commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -271,22 +274,30 @@ public override bool HandleKeyPress(KeyInput e)
 					if (!World.Selection.Actors.Any())
 						return false;
 
+					var selectedActors = World.Selection.Actors
```
This only seems to be used once? Imho you should combine the vars then:
```
var ownedActors = World.Selection.Actors
	.Where(x => !x.IsDead && eligiblePlayers.Contains(x.Owner))
	.ToList();
```

abcdefg30 reviewed (commented):
`World.RenderPlayer ?? World.LocalPlayer` not working worries me a bit, we use this pattern in several other places.

dragunoff commented:
> 
> 
> `World.RenderPlayer ?? World.LocalPlayer` not working worries me a bit, we use this pattern in several other places.

I think I can make it work. It's the line below that seems bogus. That's code that I wrote a while back when I didn't understand the engine at all. I will look into it.

dragunoff commented:
> 
> 
> > `World.RenderPlayer ?? World.LocalPlayer` not working worries me a bit, we use this pattern in several other places.
> 
> I think I can make it work. It's the line below that seems bogus. That's code that I wrote a while back when I didn't understand the engine at all. I will look into it.

Well, no :smiley: `World.RenderPlayer ?? World.LocalPlayer` doesn't cut it here. See the comments up top at https://github.com/OpenRA/OpenRA/pull/15813#pullrequestreview-196831313

I have rebased that and updated this so that double clicking on actors works the same way as using the type selection hotkey (i.e. selects for all relevant players).

I'm not willing to spend much more time on this so I'm ok to close it if it's not considered merge ready.

pchote commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -245,22 +249,24 @@ public override string GetCursor(int2 screenPos)
-			var player = World.RenderPlayer ?? World.LocalPlayer;
+			// Players to be included in the selection (the viewer or all players in "Disable shroud" / "All players" mode)
+			var viewer = (World.LocalPlayer == null || World.LocalPlayer.Spectating) ? World.RenderPlayer : World.LocalPlayer;
+			var eligiblePlayers = viewer == null || (viewer.NonCombatant && viewer.Spectating) ? World.Players : World.Players.Where(p => p == viewer);
```
Isn't `World.Players.Where(p => p == viewer)` just a confusing way of saying `new [] { viewer }`?

dragunoff replied:
Yes, pardon me, I didn't know I could do that :-) still learining the language. Fixed.

pchote commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -128,13 +128,17 @@ public override bool HandleMouseInput(MouseInput mi)
-						if (unit != null && unit.Owner == (World.RenderPlayer ?? World.LocalPlayer))
+						// Players to be included in the selection (the viewer or all players in "Disable shroud" / "All players" mode)
+						var viewer = (World.LocalPlayer == null || World.LocalPlayer.Spectating) ? World.RenderPlayer : World.LocalPlayer;
+						var eligiblePlayers = viewer == null || (viewer.NonCombatant && viewer.Spectating) ? World.Players : World.Players.Where(p => p == viewer);
```
Likewise here.

dragunoff commented:
Rebased and updated.

teinarss commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -128,13 +128,17 @@ public override bool HandleMouseInput(MouseInput mi)
-						if (unit != null && unit.Owner == (World.RenderPlayer ?? World.LocalPlayer))
+						// Players to be included in the selection (the viewer or all players in "Disable shroud" / "All players" mode)
+						var viewer = (World.LocalPlayer == null || World.LocalPlayer.Spectating) ? World.RenderPlayer : World.LocalPlayer;
+						var eligiblePlayers = viewer == null || (viewer.NonCombatant && viewer.Spectating) ? World.Players : new[] { viewer };
```
Small nitpick: You can drop the parentheses here, both these lines.

dragunoff replied:
Fixed. But only on the first line. Removing the paranthesis from the second line gives a SA1408 error.

teinarss commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -245,22 +249,24 @@ public override string GetCursor(int2 screenPos)
-			var player = World.RenderPlayer ?? World.LocalPlayer;
+			// Players to be included in the selection (the viewer or all players in "Disable shroud" / "All players" mode)
+			var viewer = (World.LocalPlayer == null || World.LocalPlayer.Spectating) ? World.RenderPlayer : World.LocalPlayer;
+			var eligiblePlayers = viewer == null || (viewer.NonCombatant && viewer.Spectating) ? World.Players : new[] { viewer };
```
Same here

dragunoff commented:
Updated to address style nit.

teinarss reviewed (dismissed):
LGTM

abcdefg30 commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -128,13 +128,17 @@ public override bool HandleMouseInput(MouseInput mi)
-						if (unit != null && unit.Owner == (World.RenderPlayer ?? World.LocalPlayer))
+						// Players to be included in the selection (the viewer or all players in "Disable shroud" / "All players" mode)
+						var viewer = (World.LocalPlayer == null || World.LocalPlayer.Spectating) ? World.RenderPlayer : World.LocalPlayer;
+						var eligiblePlayers = viewer == null || (viewer.NonCombatant && viewer.Spectating) ? World.Players : new[] { viewer };
```
What is the `viewer.NonCombatant` check here for? In general I think this can be rewritten as
```
var viewer = World.RenderPlayer ?? World.LocalPlayer;
var eligiblePlayers = viewer == null || viewer.Spectating ? World.Players : new[] { viewer };
```
(I didn't spot any bugs with that, but maybe I oversaw something.)

dragunoff replied:
The `viewer.NonCombatant` check is there to make sure that this is the special "Everyone" player. Not having it causes a bug when `LocalPlayer` is not `null` (the player has surrendered/lost) - things under shroud get selected:
![изображение](https://user-images.githubusercontent.com/1355810/62499960-2b1d8d80-b7ed-11e9-85b0-ccb01758f3e6.png)

I think I managed to find a way to use the standard viewer check. The conditions become a little lengthy so I guess pulling them in variables makes things clearer:
```csharp
var viewer = World.RenderPlayer ?? World.LocalPlayer;
var isShroudDisabled = viewer == null || (World.RenderPlayer == null && World.LocalPlayer.Spectating);
var isEveryone = viewer.NonCombatant && viewer.Spectating;
var eligiblePlayers = isShroudDisabled || isEveryone ? World.Players : new[] { viewer };
```

dragunoff replied:
A slight change and I think it works fine now. I didn't spot any bugs.
```csharp
var isEveryone = viewer != null && viewer.NonCombatant && viewer.Spectating;
```

dragunoff commented:
Updated to use the standard viewer check: `World.RenderPlayer ?? World.LocalPlayer`.

I think this check is a little wonky, because `LocalPlayer` is `null` only for spectators that did not start the game as players. I.e. if a defeated player is spectating in "Shroud off" mode then viewer won't be `null`... That's why I have a ton of conditions after that check and also why I tried to do it with a different check initially.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/873ca6e0992b517ed0f0cc07535d0fb2805f74bf...0adc1eff59274bc68181aa57f1c2c98d64321698)
